Extend player profile API to support skin changes

NickAcPT opened the pull request:
Added code that refreshes the player's skin by sending packets with a special order, telling the client to respawn the player and re-apply the game profile

NickAcPT committed:
Extend player profile API to support skin changes

Added code that refreshes the player's skin by sending packets with a special order, telling the client to respawn the player and re-apply the game profile

mikroskeem commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,185 @@
++            }
++        }
++
++        }
```
Indent is messed up here

mikroskeem commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,185 @@
+     }
++
++    private void refreshPlayer() {
++        //Get the player's dimension
```
In my opinion whitespace after `//` would be nice, e.g `// Get the player's dimension`

mikroskeem commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,185 @@
++        //Get the player's dimension
++        int dimension = getWorld().getEnvironment().getId();
++
++        //Respawn packet
```
You can remove these comments, they are redundant.

aikar commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,177 @@
+         return this;
+     }
+ 
++    // Paper start - Provide dimension number of world
```
I rather us just make the field public instead, more consistent in style.

aikar commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,177 @@
+     public void setPlayerProfile(PlayerProfile profile) {
+         EntityPlayer self = getHandle();
+         self.setProfile(CraftPlayerProfile.asAuthlibCopy(profile));
+-        List<EntityPlayer> players = server.getServer().getPlayerList().players;
```
As I said on last PR, this PR should only need to scope itself to fixing the player seeing its own skin.

Leave this change as it was, as it was fine for OTHER Entities. 
All the logic you wrote below for other entities is re-implementing the entire Entity Tracking system which we do not want to do.

So gut all that, keep this for statement, and make refreshPlayer() focused only on it self.

aikar commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,177 @@
++
++        Location l = getLocation();
++
++        Packet pos = new PacketPlayOutPosition(l.getX(), l.getY(), l.getZ(), l.getYaw(), l.getPitch(), new HashSet<>(), 0);
```
As said on the other review, gut all this, keep the reregisterPlayer loop above

aikar commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,177 @@
++
++            } else {
++                //Just send player update
++                ((CraftPlayer) pOnline).reregisterPlayer(this.getHandle());
```
This was actually a major issue. We don't want to touch anything for the player if the the receiving player cant see the source player

aikar replied:
pOnline.canSee(this)

NickAcPT committed:
Extend player profile API to support skin changes

Added code that refreshes the player's skin by sending packets with a special order, telling the client to respawn the player and re-apply the game profile

stephan-gh commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,81 @@
+     // Paper end
+ 
++    // Paper start - Provide dimension number of world
++    public int getDimension() {
```
This looks unused now.

NickAcPT replied:
oh. I forgot to delete that. Just a sec.

NickAcPT committed:
Remove unsed method

aikar commented on Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:
```diff
@@ -0,0 +1,64 @@
++
++        Packet respawn = new PacketPlayOutRespawn(entityplayer.dimension, entityplayer.world.getDifficulty(), entityplayer.world.getWorldData().getType(), entityplayer.playerInteractManager.getGameMode());
++        Location l = getLocation();
++        Packet pos = new PacketPlayOutPosition(l.getX(), l.getY(), l.getZ(), l.getYaw(), l.getPitch(), new HashSet<>(), 0);
```
Ok going to bunch everything up into one review comment, overall this is looking a lot better now.

But lets clean up some naming/impl details.

1) remove the import up top for CraftItemStack, remnant from previous version
2) rename l to loc, we prefer to not have 1 letter variables outside of positional stuff like x/y/z
3) rename playerCon to just connection
4) inline the packets so its connection.sendPacket(new PacketPlayOut*), we don't need variables for that too.
5) Add a comment on where this logic can be found in EntityPlayer or such incase future updates change the logic and we need to update this.

I think that'll cover it on quality wise and be good to go at that point.

NickAcPT committed:
Fixed more stuff
